src/SqliteContractCache.ts

```diff
@@ -0,0 +1,245 @@
+        result.sort_key,
+        JSON.parse(result.value)
+      );
+    }
+    return null;
+  }
+
+  async put(stateCacheKey: CacheKey, value: V): Promise<void> {
+    const strVal = JSON.stringify(value);
+    this.db
+      .prepare(
+        "INSERT OR REPLACE INTO sort_key_cache (key, sort_key, value) VALUES (@key, @sort_key, @value)"
+      )
+      .run({
+        key: stateCacheKey.key,
+        sort_key: stateCacheKey.sortKey,
+        value: strVal,
+      });
+    await this.removeOldestEntries(stateCacheKey.key);
+  }
+
+  private async removeOldestEntries(key: string) {
+    this.db
+      .prepare(
+        `
+            WITH sorted_cache AS
+                     (SELECT id, row_number() over (ORDER BY sort_key DESC) AS rw
+                      FROM sort_key_cache
+                      WHERE key = ?)
+            DELETE
+            FROM sort_key_cache
+            WHERE id IN (SELECT id FROM sorted_cache WHERE rw > ?);
+        `
+      )
+      .run(key, this.sqliteCacheOptions.maxEntriesPerContract);
+  }
+
+  async delete(key: string): Promise<void> {
+    this.db.prepare("DELETE FROM sort_key_cache WHERE key = ?").run(key);
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  async open(): Promise<void> {}
+
+  async close(): Promise<void> {
+    if (this._db) {
+      this._db.close();
+    }
+  }
+
+  async begin(): Promise<void> {
+    this.db.prepare("BEGIN;");
+  }
+
+  async rollback() {
+    this.db.prepare("ROLLBACK;");
+  }
+
+  async commit() {
+    this.db.prepare("COMMIT;");
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  async dump(): Promise<any> {
+    throw new Error("Not implemented");
+  }
+
+  async getLastSortKey(): Promise<string | null> {
+    const lastSortKey = this.db
+      .prepare("SELECT max(sort_key) FROM sort_key_cache")
+      .pluck()
+      .get();
+    return lastSortKey == "" ? null : lastSortKey;
+  }
+
+  storage<S>(): S {
+    return this.db as S;
+  }
+
+  /**
+   Let's assume that given contract cache contains these sortKeys: [a, b, c, d, e, f]
+   Let's assume entriesStored = 2
+   After pruning, the cache should be left with these keys: [e,f].
+
+   const entries = await contractCache.keys({ reverse: true, limit: entriesStored }).all();
+   This would return in this case entries [f, e] (notice the "reverse: true").
+
+   await contractCache.clear({ lt: entries[entries.length - 1] });
+   This effectively means: await contractCache.clear({ lt: e });
+   -> hence the entries [a,b,c,d] are removed and left are the [e,f]
+   */
+  async prune(entriesStored = 5): Promise<PruneStats> {
+    if (!entriesStored || entriesStored <= 0) {
+      entriesStored = 1;
+    }
+
+    const allItems = this.db
+      .prepare(
+        `SELECT count(*)
+       FROM sort_key_cache`
+      )
+      .pluck()
+      .get();
+    const result = this.db
+      .prepare(
+        `
+            WITH sorted_cache AS
```

Comment:
and this works per contract, right? i.e. partition by "key" guarantees that, right?

Reply:
yes, this query will group items by key (in this case contract id), order by sortKey desc and assign row_number